pkg/pub_package_reader/lib/src/pubspec_content_override.dart

```diff
@@ -0,0 +1,228 @@
+  }
+
+  try {
+    return _fixupBrokenVersionAndConstraints(pubspecYaml);
+  } on FormatException catch (e, st) {
+    _logger.info('Unable to parse pubspec.', e, st);
+  }
+  // fallback: don't override
+  return pubspecYaml;
+}
+
+/// Fixes broken version and version constraints in [pubspecYaml] and returns a new YAML string.
+///
+/// Prior to [pub_semver#63] version numbers and version constraints were accepted with any
+/// character as separator, instead of requiring  `.` (dot). This occurred because the regular expression
+/// in `package:pub_semver` didn't escape dots as `\.`, but just contained `.` (dot, meaning any character). (see https://github.com/dart-lang/pub_semver/commit/cee044a3dc8)
+///
+/// Thus, package versions published prior to 2022-01-01, may contain _version numbers_,
+/// _version constraints_ and _SDK constraints_ that don't use `.` (dot) as separator. As removing
+/// these _package versions_ would break existing users, this function will rewrite the version numbers
+/// and version constraints to have a valid format equivalent to how older versions of `pub_semver`
+/// would have interpreted the version number / version constraints.
+///
+/// [pub_semver#63]: https://github.com/dart-lang/pub_semver/pull/63
+String _fixupBrokenVersionAndConstraints(String pubspecYaml) {
+  final root = yaml.loadYaml(pubspecYaml);
+  if (root is! Map) {
+    _logger.warning(
+        'Unable to parse YAML for package, in _fixupBrokenVersionAndConstraints!');
+    return pubspecYaml; // return original as no changes can be made
+  }
+
+  // sanity check on the name again, now with parsed yaml
+  final name = root['name'];
+  if (name is! String || !_packagesWithBadVersions.contains(name)) {
+    return pubspecYaml; // return original as no changes needed
+  }
+
+  final editor = YamlEditor(pubspecYaml);
+  var hasBeenUpdated = false;
+  for (final c in _detectVersionEditCandidates(root)) {
+    final updated = c.value
+        // Version constraints may have more than one version expression, separated by space.
+        // TODO: consider replacing all whitespaces to spaces.
+        .split(' ')
+        // Each part of that expression may be mapped by the matcher only full,
+        // otherwise no replacement will be applied.
+        .map((part) => part.replaceFirstMapped(_lenientRegExp, (match) {
+              final range = match[1] ?? '';
+              final major = int.parse(match[2]!);
+              final minor = int.parse(match[3]!);
+              final patch = int.parse(match[4]!);
+              final pre = match[6];
+              final build = match[9];
+              final versionStr =
+                  Version(major, minor, patch, pre: pre, build: build)
+                      .toString();
+              return '$range$versionStr';
+            }))
+        // Joining the parts by the same separator character.
+        .join(' ');
+    if (updated != c.value) {
+      try {
+        VersionConstraint.parse(updated);
```

Comment:
Some candidates are versions and some are version constraints.

Couldn't we add this as a boolean on `_VersionEditCandidate` so that we check against `Version.parse` for versions?